skyd/skyd.go

```diff
@@ -1,16 +1,137 @@
 package skyd
 
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"gitlab.com/NebulousLabs/errors"
+	"gitlab.com/SkynetLabs/skyd/node/api/client"
+	"gitlab.com/SkynetLabs/skyd/skymodules"
+)
+
+var (
+	// skylinksCache is a local cache of the list of skylinks pinned by the
+	// local skyd
+	skylinksCache pinnedSkylinksCache
+	// skylinksCacheDuration defines the duration of skylinksCache
+	skylinksCacheDuration = time.Hour
+)
+
 type (
 	// Client allows us to call the local skyd instance.
 	Client struct {
+		staticClient *client.Client
+	}
+	// ClientInterface describes the interface exposed by Client, so we can
+	// mock it for testing.
+	ClientInterface interface {
+		Pin(skylink string) error
+		PinnedSkylinks() (skylinks []string, err error)
+		Unpin(skylink string) error
+	}
+
+	// renterDirCache is a simple cache of the renter's directory information,
+	// so we don't need to fetch that for each skylink we potentially want to
+	// pin/unpin.
+	pinnedSkylinksCache struct {
+		Skylinks   []string
+		Expiration time.Time
+		mu         sync.Mutex
 	}
 )
 
-/*
-TODO
- - env vars for talking to skyd - host, port, apipassword, etc.
- - methods for calling skyd:
-	- list all pinned skylinks
-	- pin a skylink
-	- unpin a skylink
-*/
+// NewClient creates a new skyd client.
+func NewClient(host, port, password string) *Client {
+	opts := client.Options{
+		Address:       fmt.Sprintf("%s:%s", host, port),
+		Password:      password,
+		UserAgent:     "Sia-Agent",
+		CheckRedirect: nil,
+	}
+	return &Client{
+		staticClient: client.New(opts),
+	}
+}
+
+// Pin instructs the local skyd to pin the given skylink.
+func (c *Client) Pin(skylink string) error {
+	pinned, err := c.isPinned(skylink)
+	if err != nil {
+		return err
+	}
+	if pinned {
+		// The skylink is already locally pinned, nothing to do.
+		return nil
+	}
+	sp, err := skymodules.NewSiaPath(skylink)
+	if err != nil {
+		return errors.AddContext(err, "failed to create siapath")
+	}
+	spp := skymodules.SkyfilePinParameters{
+		SiaPath: sp,
+	}
+	return c.staticClient.SkynetSkylinkPinPost(skylink, spp)
```

Comment:
Changing the code here to use a random sia path is really straightforward, e.g.
```go
	spp := skymodules.SkyfilePinParameters{
		SiaPath: skymodules.RandomSiaPath(),
	}
	return c.staticClient.SkynetSkylinkPinPost(skylink, spp)
```
Why wouldn't we want to do that and get immediately unblocked, rather than waiting for changes in `skyd`? There's nothing preventing us from making those changes, anyway.

Reply:
A follow up isn't blocking :-) just a reminder to clean something up. We also just merged `lazypin` so the follow up could be just to update this to lazypin when we are ready and that client method already doesn't accept a siapath.